cpp/src/examples/llamacpp/llamacpp_handler.cc

```diff
@@ -0,0 +1,301 @@
+                            torchserve::PayloadType::kDATA_TYPE_STRING,
+                            "runtime_error, failed to load tensor");
+    } catch (const c10::Error& e) {
+      TS_LOGF(ERROR, "Failed to load tensor for request id: {}, c10 error: {}",
+              request.request_id, e.msg());
+      auto response = (*response_batch)[request.request_id];
+      response->SetResponse(500, "data_type",
+                            torchserve::PayloadType::kDATA_TYPE_STRING,
+                            "c10 error, failed to load tensor");
+    }
+  }
+
+  return batch_ivalue;
+}
+
+c10::IValue LlamaCppHandler::Inference(
+    std::shared_ptr<void> model, c10::IValue& inputs,
+    std::shared_ptr<torch::Device>& device,
+    std::pair<std::string&, std::map<uint8_t, std::string>&>& idx_to_req_id,
+    std::shared_ptr<torchserve::InferenceResponseBatch>& response_batch) {
+  torch::InferenceMode guard;
+  auto batch_output_vector = c10::impl::GenericList(torch::TensorType::get());
+  try {
+    for (const auto input : inputs.toTensorList()) {
+      torch::Tensor tokens_list_tensor = input.get().toTensor();
+
+      int64_t num_elements = tokens_list_tensor.numel();
+
+      int64_t* data_ptr = tokens_list_tensor.data_ptr<int64_t>();
+      std::vector<llama_token> tokens_list;
+
+      for (int64_t i = 0; i < num_elements; ++i) {
+        tokens_list.push_back(data_ptr[i]);
+      }
+      const int n_gen = std::min(32, max_context_size);
+
+      std::vector<torch::Tensor> tensor_vector;
+
+      long pos = 0;
+      while (pos < n_gen) {
+        // evaluate the transformer
+
+        int n_past = pos == 0 ? 0 : llama_get_kv_cache_token_count(llama_ctx);
+
+        if (llama_eval(llama_ctx, tokens_list.data(), int(tokens_list.size()),
+                       n_past)) {
+          std::cout << "Failed to eval\n" << __func__ << std::endl;
+          break;
+        }
+
+        tokens_list.clear();
+
+        // sample the next token
+
+        llama_token new_token_id = 0;
+
+        auto logits = llama_get_logits(llama_ctx);
+        auto n_vocab = llama_n_vocab(llamamodel);
+
+        std::vector<llama_token_data> candidates;
+        candidates.reserve(n_vocab);
+
+        for (llama_token token_id = 0; token_id < n_vocab; token_id++) {
+          candidates.emplace_back(
+              llama_token_data{token_id, logits[token_id], 0.0f});
+        }
+
+        llama_token_data_array candidates_p = {candidates.data(),
+                                               candidates.size(), false};
+
+        new_token_id = llama_sample_token_greedy(llama_ctx, &candidates_p);
+
+        // is it an end of stream ?
+        if (new_token_id == llama_token_eos(llamamodel)) {
+          std::cout << "Reached [end of text]\n";
+          break;
+        }
+
+        // print the new token :
+        std::cout << "New Token: "
+                  << llama_token_to_piece(llama_ctx, new_token_id) << std::endl;
```

Comment:
do we need log each output?

Reply:
Not necessary, its an example so I though about keeping it, but can log as debug instead